File: kotlin/src/main/kotlin/de/florian/adventofcode/y2021/Day01.kt
package de.florian.adventofcode.y2021

import de.florian.adventofcode.AoCDay

fun main() { Day01().exec() }

class Day01 : AoCDay() {
    private fun convertInput(input: String) = input.split("\n").map { Integer.parseInt(it.trim()) }

    private val numbers = convertInput(Inputs_2021.DAY_01)

    private fun measurementIncreased(measurements: List<Int>, windowSize: Int = 1, stepSize: Int = 1): Int {
        return measurements
            .windowed(windowSize, stepSize){it.sum()}
            .zipWithNext()
            .count { (a,b) -> a < b }
    }

    override fun part1(): String {
        return "${measurementIncreased(numbers)}"
    }

    override fun part2(): String {
        return "${measurementIncreased(numbers, 3)}"
    }
}

File: typescript/src/2017/day01.ts
import {AocDay, AocDayDecorator, SolutionParts} from "../aoc";

@AocDayDecorator('2017', '1')
export class Day01 extends AocDay {
	private calcCaptcha(nextDigitIndexSupplier: (idx: number, inputLength: number) => number): number {
		const inputLength = this.input.split("").length;
		let sum = 0;
		for (let i = 0; i < inputLength; i++) {
			if (this.input[i] === this.input[nextDigitIndexSupplier(i, inputLength)]) {
				sum += Number.parseInt(this.input[i], 10);
			}
		}
		return sum;
	}

	solveImpl(): SolutionParts {
		return {
			part1: this.calcCaptcha((idx, inputLength) => (idx + 1) % inputLength),
			part2: this.calcCaptcha((idx, inputLength) => (idx + (inputLength/2)) % inputLength),
		};
	}
}